File: python/src/agentrunner/claudecode/runner.py
"""Claude Code runner implementation."""

from __future__ import annotations

import asyncio
import os
from collections.abc import AsyncIterator
from typing import Any

from ..errors import (
    CancelledError,
    NonZeroExitError,
    NoResultError,
    NotFoundError,
    TimeoutError,
)
from ..types import Message, Result
from .args import build_args
from .mapping import map_message_type, map_result
from .options import ClaudeRunOptions, Logger
from .parser import parse
from .process import SpawnFn, collect_error_detail, default_spawn, log_cmd
from .types import StreamMessage
from .version import check_version


class ClaudeSession:
    """Session encapsulates a running Claude Code agent process.

    Supports ``async for msg in session`` to iterate messages,
    and ``await session.result`` to get the final result.
    """

    def __init__(
        self,
        logger: Logger | None,
        spawn_fn: SpawnFn,
        binary: str,
        prompt: str,
        options: ClaudeRunOptions,
    ) -> None:
        self._logger = logger
        self._spawn_fn = spawn_fn
        self._binary = binary
        self._prompt = prompt
        self._options = options

        self._loop = asyncio.get_running_loop()
        self._queue: asyncio.Queue[Message | None] = asyncio.Queue()
        self._result_future: asyncio.Future[Result] = self._loop.create_future()
        self._process: asyncio.subprocess.Process | None = None
        self._aborted = False
        self._timed_out = False
        self._task: asyncio.Task[None] | None = None

        # Launch the background task.
        self._task = asyncio.ensure_future(self._run_process())

    async def _run_process(self) -> None:
        args = build_args(self._prompt, self._options)
        env = {**os.environ, **self._options.env} if self._options.env else None

        log_cmd(self._logger, self._binary, args, self._options.working_dir)

        try:
            self._process = await self._spawn_fn(
                self._binary,
                *args,
                cwd=self._options.working_dir,
                env=env,
            )
        except FileNotFoundError:
            err = NotFoundError(f"failed to start {self._binary}: command not found")
            self._result_future.set_exception(err)
            await self._queue.put(None)
            return

        if self._process.stdout is None:
            err = NotFoundError(f"failed to start {self._binary}: no stdout")
            self._result_future.set_exception(err)
            await self._queue.put(None)
            return

        # Set up timeout (options.timeout is in seconds).
        timeout_handle: asyncio.TimerHandle | None = None
        if self._options.timeout is not None and self._options.timeout > 0:
            timeout_handle = self._loop.call_later(self._options.timeout, self._on_timeout)

        init_session_id = ""
        result_msg: StreamMessage | None = None
        stdout_errors: list[str] = []

        try:
            while True:
                raw_line = await self._process.stdout.readline()
                if not raw_line:
                    break

                line = raw_line.decode("utf-8", errors="replace").rstrip("\n").rstrip("\r")
                if not line:
                    continue

                if self._aborted or self._timed_out:
                    break

                try:
                    parsed = parse(line)
                except Exception:
                    stdout_errors.append(line)
                    continue

                if parsed.type == "system" and parsed.subtype == "init" and parsed.session_id:
                    init_session_id = parsed.session_id
                if parsed.type == "result":
                    result_msg = parsed

                msg = Message(type=map_message_type(parsed.type, line), raw=line)

                await self._queue.put(msg)

            # Wait for process to finish.
            await self._process.wait()

            if timeout_handle:
                timeout_handle.cancel()

            if self._timed_out:
                self._result_future.set_exception(TimeoutError("execution timed out"))
                return

            if self._aborted:
                self._result_future.set_exception(CancelledError("execution cancelled"))
                return

            if result_msg:
                self._result_future.set_result(map_result(result_msg, init_session_id))
                return

            exit_code = self._process.returncode
            if exit_code is not None and exit_code != 0:
                stderr_bytes = await self._process.stderr.read() if self._process.stderr else b""
                stderr = stderr_bytes.decode("utf-8", errors="replace")
                detail = collect_error_detail(stderr, stdout_errors)
                if self._logger:
                    self._logger.error(
                        "CLI command failed",
                        extra={
                            "exit_code": exit_code,
                            "stderr": stderr.strip(),
                            "stdout_errors": stdout_errors,
                        },
                    )
                self._result_future.set_exception(
                    NonZeroExitError(exit_code, f"exit {exit_code}: {detail}")
                )
                return

            self._result_future.set_exception(NoResultError())
        finally:
            await self._queue.put(None)
            if self._process.returncode is None:
                try:
                    self._process.kill()
                except ProcessLookupError:
                    pass
            if timeout_handle:
                timeout_handle.cancel()

    def _on_timeout(self) -> None:
        self._timed_out = True
        if self._process and self._process.returncode is None:
            try:
                self._process.kill()
            except ProcessLookupError:
                pass

    def __aiter__(self) -> AsyncIterator[Message]:
        return self._message_iter()

    async def _message_iter(self) -> AsyncIterator[Message]:
        while True:
            msg = await self._queue.get()
            if msg is None:
                break
            yield msg

    @property
    def result(self) -> asyncio.Future[Result]:
        return self._result_future

    def abort(self) -> None:
        self._aborted = True
        if self._process and self._process.returncode is None:
            try:
                self._process.kill()
            except ProcessLookupError:
                pass

    def send(self, input: Any) -> None:
        raise NotImplementedError("send is not yet supported")


class ClaudeRunner:
    """Claude Code runner.

    Construct directly with keyword arguments::

        runner = ClaudeRunner(binary="/usr/local/bin/claude", logger=my_logger)

    For testing, pass ``_spawn`` to inject a fake subprocess spawner::

        runner = ClaudeRunner(_spawn=my_fake_spawn)
    """

    def __init__(
        self,
        *,
        binary: str = "claude",
        logger: Logger | None = None,
        _spawn: SpawnFn | None = None,
    ) -> None:
        self._binary = binary
        self._logger = logger
        self._spawn_fn: SpawnFn = _spawn or default_spawn
        self._has_custom_spawn = _spawn is not None
        self._version_checked = False

    async def _ensure_version(self) -> None:
        """Check the CLI version once per runner instance."""
        if self._version_checked or self._has_custom_spawn:
            return
        await check_version(self._binary)
        self._version_checked = True

    def start(
        self,
        prompt: str,
        options: ClaudeRunOptions | None = None,
    ) -> ClaudeSession:
        return ClaudeSession(
            self._logger,
            self._spawn_fn,
            self._binary,
            prompt,
            options or ClaudeRunOptions(),
        )

    async def run(
        self,
        prompt: str,
        options: ClaudeRunOptions | None = None,
    ) -> Result:
        await self._ensure_version()
        session = self.start(prompt, options)
        async for _msg in session:
            pass
        return await session.result

    async def run_stream(
        self,
        prompt: str,
        options: ClaudeRunOptions | None = None,
    ) -> ClaudeSession:
        """Start a session and return it for streaming.

        The returned session is async-iterable and also provides
        ``session.result`` and ``session.abort()``.
        """
        await self._ensure_version()
        return self.start(prompt, options)

File: agentrunner/claudecode/runner.go
// Package claudecode provides a Runner implementation for invoking Claude Code CLI
// programmatically. It implements the common Runner interface using claude -p with
// stream-json output format.
package claudecode

import (
	"bufio"
	"bytes"
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/driangle/agent-runner/agentrunner"
)

// Compile-time interface assertion.
var _ agentrunner.Runner = (*Runner)(nil)

// MinCLIVersion is the minimum supported Claude Code CLI version.
const MinCLIVersion = "1.0.12"

// commandBuilder creates an *exec.Cmd for the given binary and arguments.
// Inject a custom builder in tests to avoid spawning a real CLI process.
type commandBuilder func(ctx context.Context, name string, args ...string) *exec.Cmd

// RunnerOption configures a Runner.
type RunnerOption func(*Runner)

// WithBinary overrides the CLI binary name (default "claude").
func WithBinary(name string) RunnerOption {
	return func(r *Runner) { r.binary = name }
}

// withCommandBuilder injects a custom command builder for testing.
func withCommandBuilder(cb commandBuilder) RunnerOption {
	return func(r *Runner) { r.cmdBuilder = cb }
}

// WithLogger sets a structured logger for debug output (e.g. command args).
// When nil (the default), no logging is performed.
func WithLogger(l *slog.Logger) RunnerOption {
	return func(r *Runner) { r.logger = l }
}

// Runner implements agentrunner.Runner for Claude Code CLI.
type Runner struct {
	binary       string
	cmdBuilder   commandBuilder
	logger       *slog.Logger
	versionOnce  sync.Once
	versionError error
}

// NewRunner creates a Runner with the given options.
func NewRunner(opts ...RunnerOption) *Runner {
	r := &Runner{binary: "claude"}
	for _, o := range opts {
		o(r)
	}
	return r
}

// checkVersion verifies the CLI version is within the supported range.
// It runs once per Runner instance and caches the result.
func (r *Runner) checkVersion() error {
	// Skip version check when using a custom command builder (tests).
	if r.cmdBuilder != nil {
		return nil
	}
	r.versionOnce.Do(func() {
		out, err := exec.Command(r.binary, "--version").Output()
		if err != nil {
			// Don't fail hard — the binary may not be found, which Start handles.
			return
		}
		version := strings.TrimSpace(string(out))
		if !isVersionAtLeast(version, MinCLIVersion) {
			r.versionError = fmt.Errorf("%w: claude CLI version %s is below minimum %s", agentrunner.ErrNotFound, version, MinCLIVersion)
		}
	})
	return r.versionError
}

// isVersionAtLeast returns true if version >= minVersion using simple
// numeric comparison of dot-separated components (e.g., "1.0.12" >= "1.0.12").
func isVersionAtLeast(version, minVersion string) bool {
	parse := func(v string) []int {
		var parts []int
		for _, s := range strings.Split(v, ".") {
			n, _ := strconv.Atoi(s)
			parts = append(parts, n)
		}
		return parts
	}
	v := parse(version)
	min := parse(minVersion)
	for i := 0; i < len(min); i++ {
		vi := 0
		if i < len(v) {
			vi = v[i]
		}
		if vi < min[i] {
			return false
		}
		if vi > min[i] {
			return true
		}
	}
	return true
}

// Start launches a Claude Code CLI process and returns a Session for full
// control over the lifecycle. Pre-flight errors (version check, binary lookup,
// process start) are returned immediately. Messages arrive on session.Messages;
// the final result is available via session.Result().
func (r *Runner) Start(ctx context.Context, prompt string, opts ...agentrunner.Option) (*agentrunner.Session, error) {
	var options agentrunner.Options
	for _, o := range opts {
		o(&options)
	}

	var timeoutCancel context.CancelFunc
	if options.Timeout > 0 {
		ctx, timeoutCancel = context.WithTimeout(ctx, options.Timeout)
	}

	ctx, sessionCancel := context.WithCancel(ctx)

	cleanup := func() {
		sessionCancel()
		if timeoutCancel != nil {
			timeoutCancel()
		}
	}

	// Pre-flight: version check.
	if err := r.checkVersion(); err != nil {
		cleanup()
		return nil, err
	}

	args := buildArgs(prompt, &options)

	cmdBuilder := r.cmdBuilder
	if cmdBuilder == nil {
		if _, err := exec.LookPath(r.binary); err != nil {
			cleanup()
			return nil, fmt.Errorf("%w: %s", agentrunner.ErrNotFound, r.binary)
		}
		cmdBuilder = exec.CommandContext
	}

	cmd := cmdBuilder(ctx, r.binary, args...)
	cmd.Dir = options.WorkingDir

	if len(options.Env) > 0 {
		cmd.Env = cmd.Environ()
		for k, v := range options.Env {
			cmd.Env = append(cmd.Env, k+"="+v)
		}
	}

	r.logCmd(ctx, cmd)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		cleanup()
		return nil, fmt.Errorf("stdout pipe: %w", err)
	}

	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	if err := cmd.Start(); err != nil {
		cleanup()
		if errors.Is(err, exec.ErrNotFound) {
			return nil, fmt.Errorf("%w: %s", agentrunner.ErrNotFound, r.binary)
		}
		return nil, fmt.Errorf("start: %w", err)
	}

	onMessage := GetOnMessage(&options)

	// All pre-flight passed. Create session with the streaming goroutine.
	return agentrunner.NewSession(ctx, sessionCancel, func(ctx context.Context, msgCh chan<- agentrunner.Message) (*agentrunner.Result, error) {
		if timeoutCancel != nil {
			defer timeoutCancel()
		}

		scanner := bufio.NewScanner(stdout)
		scanner.Buffer(make([]byte, 0, 64*1024), 10*1024*1024)

		var resultMsg *StreamMessage
		var initSessionID string
		var stdoutErrors []string

		for scanner.Scan() {
			line := scanner.Bytes()
			if len(line) == 0 {
				continue
			}
			// Copy the line since scanner reuses the buffer.
			lineCopy := make([]byte, len(line))
			copy(lineCopy, line)

			parsed, parseErr := Parse(lineCopy)
			if parseErr != nil {
				stdoutErrors = append(stdoutErrors, string(lineCopy))
				continue
			}

			if parsed.Type == "system" && parsed.Subtype == "init" && parsed.SessionID != "" {
				initSessionID = parsed.SessionID
			}
			if parsed.Type == "result" {
				resultMsg = &parsed
			}

			parsedCopy := parsed
			msg := agentrunner.Message{
				Type:   mapMessageType(parsed.Type),
				Raw:    lineCopy,
				Parsed: &parsedCopy,
			}

			if onMessage != nil {
				onMessage(msg)
			}

			select {
			case msgCh <- msg:
			case <-ctx.Done():
				_ = cmd.Process.Kill()
				_ = cmd.Wait()
				if ctx.Err() == context.DeadlineExceeded {
					return nil, agentrunner.ErrTimeout
				}
				return nil, agentrunner.ErrCancelled
			}
		}

		// Check scanner error before cmd.Wait — if the scanner hit an I/O error,
		// we want to report it rather than masking it as ErrNoResult.
		scanErr := scanner.Err()

		waitErr := cmd.Wait()

		if ctx.Err() != nil {
			if ctx.Err() == context.DeadlineExceeded {
				return nil, agentrunner.ErrTimeout
			}
			return nil, agentrunner.ErrCancelled
		}

		if resultMsg != nil {
			result := mapResult(resultMsg)
			if result.SessionID == "" && initSessionID != "" {
				result.SessionID = initSessionID
			}
			return result, nil
		}

		if scanErr != nil {
			return nil, fmt.Errorf("%w: %v", agentrunner.ErrParseError, scanErr)
		}

		if waitErr != nil {
			var exitErr *exec.ExitError
			if errors.As(waitErr, &exitErr) {
				detail := collectErrorDetail(stderr.String(), stdoutErrors)
				r.logCmdFailure(ctx, exitErr.ExitCode(), stderr.String(), stdoutErrors)
				return nil, fmt.Errorf("%w: exit %d: %s", agentrunner.ErrNonZeroExit, exitErr.ExitCode(), detail)
			}
			return nil, fmt.Errorf("wait: %w", waitErr)
		}

		return nil, agentrunner.ErrNoResult
	}), nil
}

// Run executes a prompt against the Claude Code CLI and returns the final result.
// It delegates to Start, drains all messages, and returns the result.
func (r *Runner) Run(ctx context.Context, prompt string, opts ...agentrunner.Option) (*agentrunner.Result, error) {
	session, err := r.Start(ctx, prompt, opts...)
	if err != nil {
		return nil, err
	}

	// Drain all messages.
	for range session.Messages {
	}

	return session.Result()
}

// RunStream executes a prompt against the Claude Code CLI and streams parsed
// messages as they arrive. It delegates to Start and returns the session's
// message channel. The error channel receives at most one error (the session
// result error, if any) and is then closed.
func (r *Runner) RunStream(ctx context.Context, prompt string, opts ...agentrunner.Option) (<-chan agentrunner.Message, <-chan error) {
	session, err := r.Start(ctx, prompt, opts...)

	errCh := make(chan error, 1)

	if err != nil {
		errCh <- err
		close(errCh)
		ch := make(chan agentrunner.Message)
		close(ch)
		return ch, errCh
	}

	go func() {
		defer close(errCh)
		// KeepAlive prevents the GC from finalizing the session (which calls
		// Abort and cancels the context) while we are blocked on Result().
		defer runtime.KeepAlive(session)
		result, err := session.Result()
		if err != nil {
			errCh <- err
		} else if result != nil && result.IsError {
			// Not a library error — just an error result from the CLI.
			// Don't send on errCh; caller reads it from the result message.
		}
	}()

	return session.Messages, errCh
}
// mapMessageType maps a Claude stream-json type string to the common MessageType.
func mapMessageType(typ string) agentrunner.MessageType {
	switch typ {
	case "system":
		return agentrunner.MessageTypeSystem
	case "assistant":
		return agentrunner.MessageTypeAssistant
	case "user":
		return agentrunner.MessageTypeUser
	case "result":
		return agentrunner.MessageTypeResult
	case "stream_event":
		// Map to assistant since stream events carry assistant content deltas.
		return agentrunner.MessageTypeAssistant
	default:
		return agentrunner.MessageType(typ)
	}
}

func buildArgs(prompt string, opts *agentrunner.Options) []string {
	// --verbose is required with --output-format stream-json.
	args := []string{"--print", "--output-format", "stream-json", "--verbose"}

	if opts.Model != "" {
		args = append(args, "--model", opts.Model)
	}
	if opts.SystemPrompt != "" {
		args = append(args, "--system-prompt", opts.SystemPrompt)
	}
	if opts.AppendSystemPrompt != "" {
		args = append(args, "--append-system-prompt", opts.AppendSystemPrompt)
	}
	if opts.MaxTurns > 0 {
		args = append(args, "--max-turns", strconv.Itoa(opts.MaxTurns))
	}
	if opts.SkipPermissions {
		args = append(args, "--dangerously-skip-permissions")
	}

	if co := GetClaudeOptions(opts); co != nil {
		for _, t := range co.AllowedTools {
			args = append(args, "--allowedTools", t)
		}
		for _, t := range co.DisallowedTools {
			args = append(args, "--disallowedTools", t)
		}
		if co.MCPConfig != "" {
			args = append(args, "--mcp-config", co.MCPConfig)
		}
		if co.JSONSchema != "" {
			args = append(args, "--json-schema", co.JSONSchema)
		}
		if co.MaxBudgetUSD > 0 {
			args = append(args, "--max-budget-usd", strconv.FormatFloat(co.MaxBudgetUSD, 'f', -1, 64))
		}
		if co.Resume != "" {
			args = append(args, "--resume", co.Resume)
		}
		if co.Continue {
			args = append(args, "--continue")
		}
		if co.SessionID != "" {
			args = append(args, "--session-id", co.SessionID)
		}
		if co.IncludePartialMessages {
			args = append(args, "--include-partial-messages")
		}
	}

	args = append(args, "--", prompt)
	return args
}

func mapResult(msg *StreamMessage) *agentrunner.Result {
	r := &agentrunner.Result{
		Text:      msg.Result,
		IsError:   msg.IsError,
		CostUSD:   msg.TotalCostUSD,
		Duration:  time.Duration(msg.DurationMs) * time.Millisecond,
		SessionID: msg.SessionID,
	}
	if msg.Usage != nil {
		r.Usage = agentrunner.Usage{
			InputTokens:              msg.Usage.InputTokens,
			OutputTokens:             msg.Usage.OutputTokens,
			CacheCreationInputTokens: msg.Usage.CacheCreationInputTokens,
			CacheReadInputTokens:     msg.Usage.CacheReadInputTokens,
		}
	}
	return r
}

// logCmd logs the command that is about to be executed, if a logger is set.
func (r *Runner) logCmd(ctx context.Context, cmd *exec.Cmd) {
	if r.logger == nil {
		return
	}
	// Build a shell-friendly command string for easy copy-paste reproduction.
	quoted := make([]string, len(cmd.Args))
	for i, a := range cmd.Args {
		if strings.ContainsAny(a, " \t\n\"'\\") {
			quoted[i] = "'" + strings.ReplaceAll(a, "'", "'\\''") + "'"
		} else {
			quoted[i] = a
		}
	}
	r.logger.DebugContext(ctx, "executing CLI command",
		slog.String("cmd", strings.Join(quoted, " ")),
		slog.String("dir", cmd.Dir),
	)
}

// logCmdFailure logs stderr and unparseable stdout lines when the CLI exits
// with a non-zero code, if a logger is set.
func (r *Runner) logCmdFailure(ctx context.Context, exitCode int, stderr string, stdoutErrors []string) {
	if r.logger == nil {
		return
	}
	r.logger.ErrorContext(ctx, "CLI command failed",
		slog.Int("exit_code", exitCode),
		slog.String("stderr", strings.TrimSpace(stderr)),
		slog.Any("stdout_errors", stdoutErrors),
	)
}

// collectErrorDetail builds a human-readable error string from stderr and
// any unparseable stdout lines. The CLI sometimes writes errors to stdout
// as plain text rather than stream-json, so we capture both sources.
func collectErrorDetail(stderr string, stdoutErrors []string) string {
	stderr = strings.TrimSpace(stderr)
	var parts []string
	if stderr != "" {
		parts = append(parts, stderr)
	}
	if len(stdoutErrors) > 0 {
		parts = append(parts, strings.Join(stdoutErrors, "\n"))
	}
	if len(parts) == 0 {
		return "unknown error (no output from CLI)"
	}
	return strings.Join(parts, "\n")
}
